R/arc-open.R

```diff
@@ -112,3 +149,59 @@ arc_open <- function(url, token = arc_token()) {
 
   res
 }
+
+#' Test if the vector is a content url of the supplied type
+#' @keywords internal
+is_arc_content_url <- function(x, type = c("search", "item", "group", "user")) {
+  pattern <- paste0("arcgis\\.com/home/(", paste0(type, collapse = "|"), ")\\.html")
+  is.character(x) & grepl(pattern, x, perl = TRUE)
+}
+
+#' Convert a URL into a content URL
+#' @keywords internal
+arc_url_item <- function(
+    id,
+    key = "id",
+    base_url = "https://www.arcgis.com/sharing/rest/content",
+    path = "items",
+    call = caller_env()) {
+  if (is_url(id)) {
+    id <- httr2::url_parse(id)[["query"]][[key]]
+  }
+
+  check_string(id, call = call)
+
+  paste(base_url, path, id, sep = "/")
+}
+
+#' Extract the content URL type
+#' @noRd
+arc_content_url_type <- function(url) {
+  str_extract(url, "(?<=arcgis\\.com/home/)[[:lower:]]+(?=\\.html)")
+}
+
+#' Copied from stringstatic - 2024-10-31
+#' https://github.com/rossellhayes/stringstatic/blob/main/R/str_extract.R
+#' @noRd
```

Comment:
We cannot copy this code. The license is incompatible.

Reply:
Happy to take a different approach but that repo uses a CC0 license which I think is very similar to the Unlicense used by the rlang standalone scripts. Can you clarify why you think it is incompatible?

Reply:
My apologies. I mistook this for CC-BY license. Regardless, can we simplify this code 

``` r
# Simplified string extract
str_extract <- function(string, pattern) {
  matches <- regexpr(pattern, string, perl = TRUE)
  regmatches(string, matches)
}

# define the url
url <- "https://www.arcgis.com/home/item.html?id=10df2279f9684e4a9f6a7f08febac2a9"

# parse the url and grab the path don't worry about rest of url
path <- httr2::url_parse(url)$path

# regex for valid item types
valid_stuff <- "search|item|group|user"

# fetch the type
str_extract(path, valid_stuff)
#> [1] "item"
```

Reply:
Thanks! I'd been planning to do something like this and all the suggestions look good. Will get back to this soon.